docs/ADR-X-Better-call-stacks-of-io-exceptions.md

```diff
@@ -0,0 +1,79 @@
+# Status
+
+📜 Proposed 2025-03-04
+
+# Context
+
+This ADR enriches exceptions thrown from `IO` functions with call stacks up to the function usage site.
+
+## The problem
+This ADR bases on [[ADR-8-Use-RIO-in-cardano‐cli]].
+
+The goal of this ADR is to provide better call stacks in `IO` exceptions.
+The problem with exceptions thrown from IO monad is that they do not carry call stack from where they were thrown.
+This is not very convenient when for example doing multiple file access or network operations in a row.
+You won't know which one has thrown an exception: you will have only a vague exception message without many details, for example:
+```
+Network.Socket.recvBuf: resource vanished (Connection reset by peer)
+```
+
+## Solution
+
+The currently used type alias adds call stacks to the functions:
+```haskell
+type CIO e a = HasCallStack => CIO e a
+```
+We can make sure that `IO` exceptions thrown in `IO` actions are captured and rethrown with a helper `runIO` function, meant to be used instead of `liftIO`:
+
+```haskell
+import UnliftIO.Exceptions (catchAny, throwIO)
+
+runIO :: IO a -> CIO a
+runIO m = withFrozenCallStack $ catchAny (liftIO m) (throwIO . mkE)
+ where
+  mkE :: SomeException -> IoeWrapper
+  mkE e = withFrozenCallStack $ IoeWrapper e
+
+-- | An exception wrapper type, adding a call stack to it
+data IoeWrapper = HasCallStack => IoeWrapper SomeException
+
+deriving instance Show IoeWrapper
+
+instance Exception IoeWrapper where
+  displayException (IoeWrapper (SomeException e)) =
+    constructorName <> ": " <> displayException e <> "\n" <> prettyCallStack callStack
+   where
+    constructorName =  tyConName . typeRepTyCon $ typeOf e
+```
+The provided `runIO` wraps synchronous exceptions into `IoeWrapper`, which adds additional information about call stack.
+
+The resulting call stack will contain entries up to the point where `runIO` was called.
+This means that in the following code:
+```haskell
+foo :: CIO e ()
+foo = do
+  someFunction
+  runIO $ do
+    someOtherFunction
+    someExceptionThrowingFunction
+```
+the exception thrown from `someExceptionThrowingFunction` and wrapped using `runIO` will only point to the place where `runIO` was called.
```

Comment:
The fact that we only get a call-stack of the do block, and not the actual instruction that caused the exception, it is a bit unfortunate. But not sure what we can do about it. We could have a policy of only using `runIO` in a single call, but that is not great either.

Reply:
That example is a bit pathological to illustrate the problem I think we should use `runIO` on each IO action separately to have more precise call stack.